src/dataflow/src/render/upsert.rs

```diff
@@ -0,0 +1,271 @@
+
+use std::collections::HashMap;
+
+use differential_dataflow::hashable::Hashable;
+use differential_dataflow::lattice::Lattice;
+
+use timely::dataflow::channels::pact::{Exchange, Pipeline};
+use timely::dataflow::operators::generic::{operator, Operator};
+use timely::dataflow::{Scope, Stream};
+use timely::progress::Antichain;
+
+use dataflow_types::{DataEncoding, DataflowError, LinearOperator, Timestamp};
+use repr::{Datum, RelationType, Row, RowArena};
+
+use crate::decode::decode_upsert;
+use crate::operator::StreamExt;
+use crate::source::SourceOutput;
+
+/// Entrypoint to the upsert-specific transformations involved
+/// in rendering a stream that came from an upsert source.
+/// Upsert-specific operators are different from the rest of
+/// the rendering pipeline in that their input is a stream
+/// with two components instead of one, and the second component
+/// can be null or empty.
+#[allow(clippy::too_many_arguments)]
+pub fn pre_arrange_from_upsert_transforms<G>(
+    stream: &Stream<G, SourceOutput<Vec<u8>, Vec<u8>>>,
+    encoding: DataEncoding,
+    key_encoding: DataEncoding,
+    debug_name: &str,
+    worker_index: usize,
+    as_of_frontier: Antichain<Timestamp>,
+    linear_operator: &mut Option<LinearOperator>,
+    src_type: &RelationType,
+) -> (
+    Stream<G, (Row, Option<Row>, Timestamp)>,
+    Stream<G, DataflowError>,
+)
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    // This operator changes the timestamp from capability to message payload,
+    // and applies `as_of` frontier compaction. The compaction is important as
+    // downstream upsert preparation can compact away updates for the same keys
+    // at the same times, and by advancing times we make more of them the same.
+    let stream = stream.unary(Pipeline, "AppendTimestamp", move |_, _| {
+        let mut vector = Vec::new();
+        move |input, output| {
+            input.for_each(|cap, data| {
+                data.swap(&mut vector);
+                let mut time = cap.time().clone();
+                time.advance_by(as_of_frontier.borrow());
+                output
+                    .session(&cap)
+                    .give_iterator(vector.drain(..).map(|x| (x, time.clone())));
+            });
+        }
+    });
+
+    // Deduplicate records by key
+    let deduplicated = prepare_upsert_by_max_offset(&stream);
+
+    //Decode
+    let decoded = decode_upsert(
+        &deduplicated,
+        encoding,
+        key_encoding,
+        debug_name,
+        worker_index,
+    );
+
+    if let Some(operator) = linear_operator {
+        // skip applying the linear operator when it is trivial
+        if operator.predicates.is_empty()
+            && operator.projection == (0..src_type.arity()).collect::<Vec<_>>()
+        {
+            *linear_operator = None;
+        } else {
+            return apply_linear_operators(&decoded, &operator, src_type);
+        }
+    }
+    let scope = &decoded.scope();
+    (decoded, operator::empty(scope))
```

Comment:
Maybe worth a comment, but I think there is a choice between decoding/filtering first, and deduplicating by key. I can understand the arguments for either ordering, but I (personally) would jot down the local design constraints. I *think* either order is fine, with the trade-off being that decoding costs compute and deduplication costs memory. Not sure, but if we know anything about it let's record it!